### PR TITLE
Kpi user

### PR DESCRIPTION
**Pull request by azizullahsaeidi**
1. Get top 10 active users
2. Get total registered users
3. Get total users base on gender
4. Get total users
5. Get total Google users
6. Get total Facebook users

Please run below command
`
php artisan migrate --path=/database/migrations/2024_04_30_142706_add_provider_to_users_table.php
`

### CODE REVIEW

**Review by uyscuti-wiki — CHANGES_REQUESTED** [review]
Looks good. I've left some comments to improve some aspects of the code.

File: routes/web.php [hunk 1]
```diff
@@ -286,6 +287,10 @@
         Route::get('/', 'index')->name('subscribe.index');
         Route::post('', 'store')->name('subscribe.store');
     });
+
+    // User Analytics
+    Route::get('admin/user-analytics',  [UserAnalyticsController::class, 'index'])->middleware('admin');
```
Comment (uyscuti-wiki) [inline routes/web.php:292]: Let's also unify the routes for analytics under `admin/analytics/users` and `admin/analytics/resources`.

File: .env.example
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply .env.example:1]: Removed

File: app/Http/Controllers/UserAnalyticsController.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        $totalRegisteredUsers = $this->getTotalRegisteredUsers($request);
+        $totalUsersBaseOnGenders = $this->getTotalUsersBaseOnGender($request);
+        $totalUsers = $this->getTotalUsersBaseOnRegistrationSource($request);
+        $totalGoogleUsers = $this->getTotalUsersBaseOnRegistrationSource($request, 'google');
+        $totalFacebookUsers = $this->getTotalUsersBaseOnRegistrationSource($request, 'facebook');
+
+        return view('admin.analytics.users.index', compact(['roles', 'totalUsersBaseOnGenders', 'totalRegisteredUsers', 'totalUsers', 'totalGoogleUsers', 'totalFacebookUsers', 'top10ActiveUsers']));
+    }
+
+    private function getTotalUsersBaseOnRole($request): Collection
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/UserAnalyticsController.php:27]: Done

File: app/Http/Controllers/UserAnalyticsController.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private function getTotalRegisteredUsers($request): float
+    {
+        return User::when($request->date_from && $request->date_to, function ($query) use ($request) {
+            $query->whereBetween('created_at', [$request->date_from, $request->date_to]);
+        })->count();
+    }
+
+    private function getTotalUsersBaseOnRegistrationSource($request, $providerName = null): float
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/UserAnalyticsController.php:56]: Renamed

File: app/Http/Controllers/UserAnalyticsController.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        return Role::withCount([
+            'users' => function ($query) use ($request) {
+                if ($request->date_from && $request->date_to) {
+                    $query->whereBetween('created_at', [$request->date_from, $request->date_to]);
+                }
+            },
+        ])->get();
+    }
+
+    private function getTotalUsersBaseOnGender($request): Collection
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/UserAnalyticsController.php:38]: Done

File: resources/views/admin/analytics/users/index.blade.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
+@extends('admin.layout')
+@section('admin.content')
+    <div class="content-wrapper">
+        <div class="container-fluid">
+            <div class="pb-4">
+                <form method="get" action="{{ url('admin/user-analytics') }}">
+                    @csrf
+                    <div class="row">
+                        <div class="col-md-2">
+                            <label>From Date</label>
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply resources/views/admin/analytics/users/index.blade.php:10]: Done

File: resources/views/admin/analytics/users/index.blade.php [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                            </div>
+                        </div>
+
+                        {{-- Top 10 active users  --}}
+                        <div class="col-sm-6 col-md-4 col-lg-3">
+                            <div class="card border-secondary mb-3">
+                                <div class="card-header">Top 10 active users </div>
+                                <div class="card-body text-secondary p-2">
+                                    <div class="card-text">
+                                        @forelse ($top10ActiveUsers as $top10ActiveUser)
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply resources/views/admin/analytics/users/index.blade.php:106]: Done

File: resources/views/admin/analytics/users/index.blade.php [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                                            <div class="d-flex justify-content-between mb-2 rounded bg-light text-dark">
+                                                <div class="p-1 text-capitalize">
+                                                    {{ $loop->iteration }}.
+                                                    {{ $top10ActiveUser->first_name }} {{ $top10ActiveUser->last_name }}
+                                                </div>
+                                                <div class="p-1">
+                                                    <span class="badge badge-info">
+                                                        {{ number_format($top10ActiveUser->activity_count) }}
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply resources/views/admin/analytics/users/index.blade.php:114]: Done

File: config/services.php [hunk 8]
```diff
@@ -50,7 +50,7 @@
     'google' => [
         'client_id' => env('GOOGLE_CLIENT_ID'),
         'client_secret' => env('GOOGLE_CLIENT_SECRET'),
-        'redirect' => 'https://library.darakhtdanesh.org/login/google/callback/',
+        'redirect' => env('GOOGLE_REDIRECT', 'https://library.darakhtdanesh.org/login/google/callback/'),
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply config/services.php:53]: We have default value for it. `'redirect' => 'https://library.darakhtdanesh.org/login/google/callback/',`

File: resources/views/admin/analytics/users/index.blade.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                                        @endforelse
+                                    </div>
+                                </div>
+                            </div>
+                        </div>
+
+                        {{-- List the total admins/managers/translators.  --}}
+                        <div class="col-sm-6 col-md-4 col-lg-3">
+                            <div class="card border-secondary mb-3">
+                                <div class="card-header">List the total admins/managers/translators. </div>
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply resources/views/admin/analytics/users/index.blade.php:103]: Done

File: routes/web.php (hunk 1 above)
Comment (azizullahsaeidi) [reply routes/web.php]: Thank you, done.

File: app/Http/Controllers/UserAnalyticsController.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        $totalUsers = $this->getTotalUsersBaseOnRegistrationSource($request);
+        $totalGoogleUsers = $this->getTotalUsersBaseOnRegistrationSource($request, 'google');
+        $totalFacebookUsers = $this->getTotalUsersBaseOnRegistrationSource($request, 'facebook');
+
+        return view('admin.analytics.users.index', compact(['roles', 'totalUsersBaseOnGenders', 'totalRegisteredUsers', 'totalUsers', 'totalGoogleUsers', 'totalFacebookUsers', 'top10ActiveUsers']));
+    }
+
+    private function getTotalUsersBaseOnRole($request): Collection
+    {
+        return Role::withCount([
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/UserAnalyticsController.php]: Done

File: resources/views/admin/analytics/users/index.blade.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (uyscuti-wiki) [reply resources/views/admin/analytics/users/index.blade.php]: @azizullahsaeidi I meant just the `activity_count`, not the entire dataset.

File: resources/views/admin/analytics/users/index.blade.php [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                                                <span class="badge badge-info">
+                                                    {{ number_format($totalRegisteredUsers) }}
+                                                </span>
+                                            </div>
+                                        </div>
+                                    </div>
+                                </div>
+                            </div>
+                            <div class="card border-secondary mb-3">
+                                <div class="card-header">Total users base on registration source </div>
```
[the remark this answers is not in the document]
Comment (uyscuti-wiki) [reply resources/views/admin/analytics/users/index.blade.php:52]: @azizullahsaeidi please fix this.

File: resources/views/admin/analytics/users/index.blade.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (uyscuti-wiki) [reply resources/views/admin/analytics/users/index.blade.php]: We need the top 10 active users data, just not the numbers against them.

File: resources/views/admin/analytics/users/index.blade.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply resources/views/admin/analytics/users/index.blade.php]: Done

**Review by uyscuti-wiki — APPROVED** [review]
Looks great, thank you Aziz!

**Pull request merged**